### ISSUE TITLE
[Request] Implement compiler delay on file change

### ISSUE THREAD

**Issue by jmhickman** [Ackee-Blockchain/tools-for-solidity-vscode#36]
Great project, been enjoying using it.

I've noticed that as I add comments to code, I seem to be queuing up re-compiles. The number of times the file gets re-compiled seems to scale with the number of words/keypresses/changes I make to the file, and it can add up to half a minute or more on my system before it 'settles down.'

It seems like a delay should be implemented (500ms, 2000ms, etc) so that the recompile cycles occur less frequently?

**Comment by jmhickman** [comment]
Just wanted to bump this particular issue for attention. When projects take tens of seconds to fully compile (even with optimization runs disabled), these queued up compile passes can really drag on productivity. While the `woke` process is doing its thing, you lose most of the codelens features and peek-on-mouseover for types and such. 

Really wanting a sensible delay parameter or even just a "only re-compile on hotkey" command.

**Comment by michprev** [comment]
Hello,
thank you for your interest in our extension.

Compilation is a challenging topic, especially in the context of the LSP server. I am assuming you are using the extension with Foundry projects. Before we start digging deeper/start discussing the best solution, could you please try to add `lib` into the `woke.compiler.solc.ignore_paths` VS Code config? Does the situation get any better?

**Comment by jmhickman** [comment]
This is my current `woke.compiler.solc.ignore_paths`:

![image](https://user-images.githubusercontent.com/6581761/211604962-0c016ed7-b5d4-4dbd-808f-859cf3d2064e.png)

I've tried to pare it down as much as I can.

This occurs with hardhat-based projects and Foundry ones.

For this specific issue, it's more about how writing anything into the file immediately triggers a recompile, even when I'm still writing. The number of times I see "compiling: x/y" in the status bar seems to be proportional to the amount of changes I make in the file. The size of a compilation cycle makes the problem go from "not too bad" to "this makes me wait tens of seconds for just writing a long comment."

Other extensions seem to deal with this by having a setting that delays any automated action by a certain number of milliseconds.

There seems to be a related issue with the `woke` process crashing, I posted an example of the traces I'm seeing in the woke repo.

**Comment by michprev** [comment]
I see.

The current compilation process is quite complicated. Unless you have compilation errors in your project, you should be experiencing two compilation runs (on average) in the following scenario:
- you start typing, the first compilation run begins
- while you type, more requests from the LSP client are being received - we cannot add them to the already running compilation run
- you stop typing
- the first compilation finishes
- the second compilation run starts to process all the changes that we received during the first compilation

A single compilation run can be recognized by  "compiling: x/y" appearing and disappearing. "y" in the text refers to something we internally call compilation units which ten equals the number of independent `solc` instances started in a single compilation run.

We always process all changes that arrive in our queue up to a given time, no matter how long we collect them (i.e., how long the previous compilation runs). That is why you should be experiencing up to double the compilation time after you stop typing - the currently running compilation must finish. Then another must start to process all the remaining changes. This is why I am not sure adding a delay will greatly impact current waiting times. On the other hand, it might introduce unnecessary delays for projects where we can compile fast.

Also, we added a new feature in the latest release - we can provide LSP analysis even if we don't have the latest compilation build - we can use the last successful build. However, due to the nature of Python, we cannot handle LSP requests and perform compilation simultaneously (Python plays poorly with parallelism). But we can respond to LSP requests while waiting for all `solc` instances to finish. From my experience, this does not happen in the current version of the extension. I will investigate that.

TL;DR;
- we can add some delay, but I am not sure it will help greatly - for now, I would consider it a fallback solution
- some LSP features (Go to def, hover) should work even while compiling, but do not - I will investigate
- you should experience up to two compilation runs after you stop typing
    - if you experience anything other than that, I would call it a bug
- if you are experiencing this issue with a public project, it would be helpful if you could post a link here

**Comment by jmhickman** [comment]
Okay, if the LSP features should work while a current compile is running, then that would at least mitigate my pain point (I can't see hover items when it's waiting for compiles to finish). If you feel the delay wouldn't solve the issue at its core and would add a lot of complexity, that's fine. I still feel that the compiles shouldn't being until I stop typing though.

> you should experience up to two compilation runs after you stop typing
if you experience anything other than that, I would call it a bug

I definitely have _something_ that's happening where this is being broken. I just had to reload the project in VSCode because the compiles were happening continuously without end for more than 5 minutes. Long long after I had stopped typing.

The current project I'm working with is hosted here: https://github.com/code-423n4/2023-01-astaria 
I'm not certain what exactly I'm doing that is causing the infinite compile issue though, so I don't know that the specific repo will help you much.

In case it helps, the machine running the compiles is an Intel Core i7-6800K CPU @ 3.40GHz (6 cores, 12 threads), with 24GB of system memory.

**Comment by michprev** [comment]
I made several improvements and fixed some bugs in our LSP server:
- major LSP features should be working even during compilation runs (go to def, document links - imports, hover, go to type def); however, the responses may be a bit "lazy" (we have to do compilation and respond to LSP messages in the same thread/process/core)
- fixed an issue when "compiling: x/y" progress would be kept running even if compilation was already finished/terminated
- found a performance issue in the LSP compiler in the particular case of the project you posted; fixing this issue significantly reduced the processing time of build artifacts (41 s => 10 s)
- added an optional delay before we start compiling after a keystroke; it will be configurable from VS Code settings as it may vary from one person to another

All of these points will require a newer version of Woke - our LSP server. The new version of Woke should be released at the end of January / at the start of February. I'll let you know when the new version is out.

**Comment by michprev** [comment]
Hey @jmhickman,
we have just released the new version of Woke (2.0.0) and Tools for Solidity (1.5.0) that implement all of the described features.

**Comment by jmhickman** [comment]
This looks great, tested it out. Thanks so much for being responsive and making the tool better. 🙏
